plugins/mysql/mysqlembed.cpp

```diff
@@ -93,27 +93,117 @@ static void fail(const char *message)
 
 // Wrappers to MySQL structures that require corresponding releases
 
+class MySQLConnection;
+
+static __thread MySQLConnection *threadCachedConnection = nullptr;
+
+#define MAX_GLOBAL_CACHE 10
+
 class MySQLConnection : public CInterface
 {
 public:
     IMPLEMENT_IINTERFACE;
-    MySQLConnection(MYSQL *_conn) : conn(_conn)
+    MySQLConnection(MYSQL *_conn, const char *_cacheOptions, bool _threadCached, bool _globalCached) : conn(_conn), threadCached(_threadCached), globalCached(_globalCached)
     {
+        if (threadCached || globalCached)
+            cacheOptions = strdup(_cacheOptions);
+        else
+            cacheOptions = nullptr;
     }
     ~MySQLConnection()
     {
         if (conn)
-            mysql_close(conn);
+        {
+            if (threadCached || globalCached)
+            {
+                Owned<MySQLConnection> cacheEntry = new MySQLConnection(*this);
+                if (threadCached)
+                    setThreadCache(cacheEntry.getLink());
+                if (globalCached)
```

Comment:
Should these be exclusive?  Otherwise the same object could be used by multiple threads at once.